main/src/com/google/refine/expr/functions/Type.java

```diff
@@ -70,7 +72,7 @@ public Object call(Properties bindings, Object[] args) {
 
     @Override
     public String getDescription() {
-        return "Returns the type of o";
+        return "Returns the type of o as a string ('string', 'date', 'number', 'array', 'boolean', 'error' or a class name)";
```

Comment:
```suggestion
        return "Returns the type of o as a string ('string', 'date', 'number', 'array', or 'boolean'). The string 'error' is returned if the value can't be evaluated";
```
Now that the bug is fixed, can it ever return a Java class name? I'd be tempted to just pare this down to simply "Returns the type of o as a string."

Reply:
> Now that the bug is fixed, can it ever return a Java class name? I'd be tempted to just pare this down to simply "Returns the type of o as a string."

@tfmorris: the question you raise is far more complex, from what I see. In Issue « 
true.type() return « java.lang.Boolean » and not « boolean »
#2510 », you will find some discussion on that, but here's a table I did:
Expression|Result|Expected result
--- | -- | :------:
row.type()|com.google.refine.expr.WrappedRow|"object: row" 
row.record.type()| com.google.refine.expr.WrappedRow$WrappedRecord |"object: record"
cell.type()| com.google.refine.expr.WrappedCell |"object: cell"
cells.type()| com.google.refine.expr.CellTuple |"object: cells"
cell.recon.features.type()| com.google.refine.model.Recon$Features |"object: features"
row.record.cells.type()| com.google.refine.expr.WrappedRow$RecordCells |"object: cells"

Because in the code, we have this: `return v.getClass().getName()`, we can never be sure of all the value that can be returned. I guess the only thing we can be sure it's that it's going to be a string…
```
                } else if (v instanceof EvalError) {
                    return "error";
                } else {
                    return v.getClass().getName();
                }
```
I think we should close this issue, and maybe open another one on how to return a type from an object. But closing this issue, we finalize the type() function with basic datatypes.

For this particular change you suggest: « The string 'error' is returned if the value can't be evaluated"; », I wouldn't write that because I don't believe it's accurate. Imagine the case of use `type(value)`, where value contains a previous stored error. But this discussion should be in issue « type(an error) should return the string "error" like the code was designed to do but is buggy #2562 ». Do you share this opinion?

Regards, Antoine

Reply:
Marking this as resolved. As discussed elsewhere, evaluation errors happen before ever reaching this function.